onnxruntime/core/providers/cpu/tensor/cast_op.cc

```diff
@@ -1,193 +1,231 @@
-  auto out_data = out.MutableData<MLFloat16>();
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-  auto in_vector = ConstEigenVectorMap<float>(in.Data<float>(), shape_size);
-  auto output_vector = EigenVectorMap<Eigen::half>(static_cast<Eigen::half*>(static_cast<void*>(out_data)),
-                                                   shape_size);
-  output_vector = in_vector.template cast<Eigen::half>();
-}
+using EnabledSrcTypes = ORT_OP_KERNEL_ARG_ENABLED_TYPE_LIST(Cast, Input, 0);
+using EnabledDstTypes = ORT_OP_KERNEL_ARG_ENABLED_TYPE_LIST(Cast, Output, 0);
 
-template <>
-inline void CastData<MLFloat16, float>(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  auto out_data = out.MutableData<float>();
-  auto in_data = in.Data<MLFloat16>();
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-#if defined(_M_AMD64)
-  MlasConvertHalfToFloatBuffer(&in_data[0].val, out_data, shape_size);
-#else
-  auto in_vector = ConstEigenVectorMap<Eigen::half>(static_cast<const Eigen::half*>(static_cast<const void*>(in_data)),
-                                                    shape_size);
-  auto output_vector = EigenVectorMap<float>(out_data, shape_size);
-  output_vector = in_vector.template cast<float>();
-#endif
-}
+using IndirectCastTypes = TypeList<MLFloat16, BFloat16>;
 
-template <>
-inline void CastData<float, BFloat16>(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  auto out_data = out.template MutableData<BFloat16>();
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-  auto in_vector = ConstEigenVectorMap<float>(in.template Data<float>(), shape_size);
-  auto output_vector = EigenVectorMap<BFloat16>(out_data, shape_size);
-  output_vector = in_vector.template cast<BFloat16>();
-}
+template <typename Type>
+using IsDirectCastType = mp_not<mp_contains<IndirectCastTypes, Type>>;
 
-template <>
-inline void CastData<BFloat16, float>(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  auto out_data = out.template MutableData<float>();
-  auto in_data = in.template Data<BFloat16>();
-  ptrdiff_t shape_size = gsl::narrow<ptrdiff_t>(shape.Size());
-  auto in_vector = ConstEigenVectorMap<BFloat16>(in_data, shape_size);
-  auto output_vector = EigenVectorMap<float>(out_data, shape_size);
-  output_vector = in_vector.unaryExpr([](BFloat16 val) { return val.ToFloat(); });
-}
-#endif
+template <typename... Types>
+using AreAllDirectCastTypes = mp_all<IsDirectCastType<Types>...>;
 
-#ifdef CAST_STRING_ENABLED
+// string cast helpers
 
 // handle floating point input separately
 template <typename SrcType>
 typename std::enable_if<std::is_floating_point<SrcType>::value, void>::type
-CastToStringData(const Tensor& in, Tensor& out, const TensorShape& shape) {
-  const int64_t len = shape.Size();
-  const auto input_data = in.DataAsSpan<SrcType>();
-  auto output_data = out.MutableDataAsSpan<std::string>();
-
-  for (int i = 0; i < len; ++i) {
```

Comment:
Wondering if there's any potential perf impact from moving the 'for' loop away from the casting logic. Possibly not, but may need to check.